File: solutions/python3/0032.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        ans = 0
        left = 0
        right = 0

        for i in range(len(s)):
            if s[i] == '(':
                left += 1
            else:
                right += 1
            if left == right:
                ans = max(ans, 2 * right)
            elif right > left:
                left = 0
                right = 0

        left = 0
        right = 0

        for i in range(len(s) - 1, -1, -1):
            if s[i] == '(':
                left += 1
            else:
                right += 1
            if left == right:
                ans = max(ans, 2 * left)
            elif left > right:
                left = 0
                right = 0

        return ans

File: solutions/cpp/0032.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int ret = 0;
        int left = 0;
        int right = 0;

        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(')
                left++;
            else
                right++;
            if (left == right)
                ret = max(ret, 2 * right);
            else if (right > left) {
                left = 0;
                right = 0;
            }
        }

        left = 0;
        right = 0;

        for (int i = s.size() - 1; i >= 0; i--) {
            if (s[i] == '(')
                left++;
            else
                right++;
            if (left == right)
                ret = max(ret, 2 * left);
            else if (left > right) {
                left = 0;
                right = 0;
            }
        }

        return ret;
    }
};